### ISSUE TITLE
Int-suite conv bias is all-zero, so a kernel that drops the bias term entirely still passes

### ISSUE THREAD

**Issue by apage224** [AmbiqAI/helia-core-tester#77]
## What is wrong

Every int-suite convolution case ships an **all-zero bias tensor**. A kernel that dropped the bias-add term entirely — or added it with the wrong sign, or to the wrong channel — would pass all of them.

`helia_core_tester/generation/ops/ConvolutionFunctions/convolve.py:96-112`:

```python
        use_bias = self.desc.get('use_bias', True)
        # A zero bias_initializer produces an all-zero bias tensor, which the
        # TFLite converter's constant-folding optimizer strips from the
        # FLOAT (non-quantized) graph entirely -- the generated CMSIS-NN
        # test then calls the kernel with a NULL bias pointer, leaving the
        # bias-add path completely untested. Use a small nonzero uniform
        # bias, deterministic from the case seed, so real bias data flows
        # through the golden and the kernel call.
        # Float cases only: the same Keras model also feeds the QUANTIZED
        # pipeline, so an ungated nonzero bias would silently perturb every
        # int8/int16 conv golden (54 cases -- caught by adversarial review).
        # Int-suite bias enrichment is a separate, deliberate change.
        _case_is_float = str(self.tensor_dtype("input", default="S8")).upper() in {"FP32", "FP16"}
        bias_initializer = (
            tf.keras.initializers.RandomUniform(minval=-0.25, maxval=0.25, seed=self.seed)
            if (use_bias and _case_is_float) else 'zeros'
        )
```

The nonzero initializer at `:110` is gated on `use_bias and _case_is_float` at `:111`. Int cases take the `'zeros'` branch. The gate is deliberate and correctly reasoned — an ungated change would have silently rewritten every quantized golden — and `:107` says so plainly:

> `# Int-suite bias enrichment is a separate, deliberate change.`

This issue is that separate change. It has not been filed until now, so it has been sitting as a comment with no tracking behind it.

## The float side is already fixed — do not re-file that

To be explicit, because this gap has a confusing history: **the float-bias problem is resolved at this pin.** An older comment thread on #53 recorded that "all 46 Convolve float cases ship all-zero bias arrays, all 10 FullyConnected float cases pass `biases = NULL`." That is no longer true for Convolve — `convolve.py:97-112` is exactly the fix, and float conv cases now carry deterministic nonzero bias. Anyone picking this up should scope it to the **int suite only** and should not reopen the float half.

(The FullyConnected float `biases = NULL` observation from that same comment was not checked by this audit and is not claimed either way here.)

## Blast radius

The comment at `:106` puts the figure at **54** int8/int16 conv goldens, measured when that change landed.

A recount at `main` `080db5f` gives **65** Convolve cases with a non-float input dtype and `use_bias` true — 56 S8 and 9 S16, out of 121 Convolve cases total. Counted by parsing every YAML document in `assets/descriptors/ConvolutionFunctions/convolve.yaml` and selecting `operator: Convolve` documents with a non-FP input dtype and `use_bias` true or absent (it defaults true).

Treat **54-65** as the affected range and confirm the exact set before scoping; the difference is most likely dtype scope or descriptor drift since the comment was written, and it does not change the shape of the problem.

**Not investigated, and worth checking before scoping:** `DepthwiseConv` and `TransposeConv` are generated by their own modules, not by `convolve.py`. Whether they have the same all-zero int bias was not established by this audit. If they do, the affected set is materially larger than the numbers above.

## Why it matters

The bias-add is not an incidental detail of a quantized convolution — it is where the per-channel offset and part of the requantization arithmetic live. With an all-zero bias:

- A kernel that skips the bias accumulate produces bit-identical output to a correct one on every one of these cases.
- Per-output-channel bias indexing is unexercised, so an off-by-one or a transposed channel index is invisible.
- The `int32` bias accumulator's overflow and saturation behavior is never approached, because nothing nonzero is ever accumulated into it.

The suite is the PR gate for heliaCORE. For this class of defect it currently reports success on every int conv case it has.

## Suggested direction

Mirror what `:109-112` already does for float, with the goldens regenerated in the same change:

1. Drop `_case_is_float` from the gate so `use_bias` alone selects the nonzero initializer, keeping the same seed-derived `RandomUniform` so the values stay deterministic.
2. Regenerate the affected int8/int16 conv goldens. This is the reason the change was deferred and it is the bulk of the work — the diff will touch every affected case's expected output, and that diff needs to be reviewed as a golden change, not skimmed.
3. Choose the bias magnitude deliberately for the quantized path rather than reusing the float `+/-0.25`. It should be large enough that dropping the bias term shifts the output past the int8 quantization step, and small enough not to saturate the accumulator. Worth measuring rather than guessing.
4. Add a mutation check proving the new cases detect a dropped bias term — the pattern established in #52 and #54, and the same discipline #76 asks for generally.

Step 3 is the part that needs judgment; steps 1 and 2 are mechanical.

Rough size: not measured. Step 2 dominates and depends on how the golden regeneration is reviewed.

## Related

- #76 — a case can be green and still not discriminate. These cases are a clean example: they pass, and they would pass just as reliably against a kernel with no bias-add at all.
- #72 — fault-case machinery for the same operator families. Independent work, but both touch the conv descriptors and the two may be cheaper to review together than separately.

Verified against `main` at `080db5f`.

**Comment by apage224** [comment]
## Scope resolved: depthwise and transpose-conv are clean, but FullyConnected is affected

The issue flagged as not investigated whether DepthwiseConv and TransposeConv share the all-zero int bias. They do not — but a family the issue does not mention does.

**Not affected.** Both set a nonzero bias unconditionally, with no float gate:

```python
# depthwise_conv.py:99-100
if dwconv_kwargs['use_bias']:
    dwconv_kwargs['bias_initializer'] = tf.keras.initializers.RandomUniform(minval=-1.0, maxval=1.0)

# transpose_conv.py:43-44
if transpose_conv_kwargs['use_bias']:
    transpose_conv_kwargs['bias_initializer'] = tf.keras.initializers.RandomUniform(
        minval=-0.5, maxval=0.5, seed=321)
```

Their int cases get real bias data, so a kernel dropping the bias term fails there.

**Affected, and not currently in this issue: FullyConnected.** `fully_connected.py:69-73` carries the identical gate to `convolve.py`:

```python
_case_is_float = str(self.tensor_dtype("input", default="S8")).upper() in {"FP32", "FP16"}
bias_initializer = (
    keras.initializers.RandomUniform(minval=-0.25, maxval=0.25, seed=self.seed)
    if (use_bias and _case_is_float) else 'zeros'
)
```

with the same rationale in the comment above it (`:67-68`): "Float cases only: the same Keras model also feeds the QUANTIZED pipeline (see convolve.py) — keep int FC goldens byte-identical."

So the scope is **Convolve and FullyConnected**, both int suites, for the same deliberate reason: changing the int bias changes the quantized goldens byte-for-byte, which is why it was deferred rather than overlooked. Whoever picks this up should plan for a golden regeneration across both families, not just conv, and should expect the diff to be large and mechanical.

This does not change the underlying point — a kernel that drops the bias term still passes those int cases — it just doubles the families involved and confirms why it was left.

Not investigated here: whether the FullyConnected `biases = NULL` observation from #53's comment history is a separate matter. It appears to be, since that concerns the float path, which is already fixed.

**Comment by apage224** [comment]
**Implementation spec (up for grabs — will self-assign if unclaimed in ~2 weeks).**

Change the int-suite conv-family generators to draw bias from a non-degenerate distribution (e.g. uniform over the int32 bias range scaled to avoid saturating the requantize) instead of zeros, and add one deliberate asymmetric-bias case per family (conv, depthwise, FC, transpose-conv) whose expected output differs from the zero-bias output on every element — so a kernel that drops the bias term fails loudly rather than passing 100% of cases. With #84 merged, expected values serialize at full precision, so no tolerance workaround is needed. Acceptance: re-run generation, show the m4/m55 int legs still pass on current heliaCORE, then mutate one kernel locally (drop its bias add) and show the suite fails.

**Comment by apage224** [comment]
Measured in #89's mutation run: with the bias term deleted from every int conv kernel path, the s4 suite kills the mutant (20/25 cases — nonzero bias) while **every s8/s16 case that executes the bias code passes** — this issue, now a number instead of an argument. The spec above stands; the fix should bring the s8/s16 suites to where s4 already is.
